### DISCUSSION TITLE
Understanding Linking and Stripping

### DISCUSSION THREAD

**Discussion by olokobayusuf** [emscripten-core/emscripten#21850] [Q&A]
Hey all, I'm trying to wrap my head around how the Emscripten toolchain handles stripping specifically around JS library symbols. My use case is simple: I'm compiling a main module that will be extended at runtime with dynamic linking (side module). The main module needs to contain `embind` symbols that the side module will try to use, but since these symbols are not used at compile-time, they will get stripped away. Here's what I've learned so far:

1. Unused symbols defined in JS library functions (e.g. `LibraryEmVal`) will be stripped as expected.
2. `wasm-ld` does not have visibility or control over JS library symbol stripping, so using `-Wl,-u,unused_symbol_name` does not work.
3. Similarly, `wasm-ld` cannot preserve `extern` symbols (i.e. symbols defined in JS), so using `-Wl,--export=unused_symbol_name` also does not work.

My question then is what part of the Emscripten toolchain handles stripping JS library functions, and what compiler flags can be used to preserve these symbols?

I've tested `-sEXPORTED_FUNCTIONS` which doesn't work because this controls making C symbols visible in JavaScript, not the other way around. Similarly, `-sEXPORTED_RUNTIME_METHODS` doesn't work because it only handles making a function available in the global `Module`, and doesn't preserve any form of stripping.

**Comment by sbc100** [comment]
If you build with `-sMAIN_MODULE` then emscripten won't strip anything and it will include basically everything by default. 

If you build with `-sMAIN_MODULE=2` then emscripten will strip dead code.  The easiest way to mark things are needed is to simply pass the side module on the command line when linking the main module.  Then any symbols needed by the side module will be automatically preserved.   If you don't have access to the side module(s) when linking the main module and you still want to use `-sMAIN_MODULE=2` then you need to add all symbols needs by the side modules(s) explicitly using `-sEXPORTED_FUNCTIONS`.

There are some docs on this here: https://emscripten.org/docs/compiling/Dynamic-Linking.html#code-size

**Reply by olokobayusuf** [reply]
@sbc100 Thanks for the quick response on this. I'm more curious to know _how this works_, versus simply using it. I'll try `-sEXPORTED_FUNCTIONS` again to confirm because I previously tried with no success.

Which parts of Emscripten handle symbol collection and retention through DCE, across both C++ and JS? Would appreciate pointers (source file, etc) just so I can get a much better understanding.

**Reply by sbc100** [reply]
There are several phases at which dead code is eliminated.   The specific process you are talking about with references across C++ and JS we call metadce.    metadce itself is a multistage process that involves the python driver code calling out to a binaryen tool and also an acorn optimizer path.   You can see the code here:  https://github.com/emscripten-core/emscripten/blob/0649692b8ab17bac09262754b68cbf6aff297963/tools/building.py#L736-L737

Are you not able to pass the side modules as input? 

Using `-sMAIN_MODULE=2` *without* passing the side modules as input is kind of pain since if involves managed potentially huge list of exported functions manually.

**Reply by olokobayusuf** [reply]
> There are several phases at which dead code is eliminated. The specific process you are talking about with references across C++ and JS we call metadce. metadce itself is a multistage process that involves the python driver code calling out to a binaryen tool and also an acorn optimizer path. You can see the code here:

This is exactly what I need, a good starting point to learn.

> Using -sMAIN_MODULE=2 without passing the side modules as input is kind of pain since if involves managed potentially huge list of exported functions manually.

We can automate this to a great degree, specifically because the symbols we need are pretty specific (e.g. `embind`). I just did a sanity check and `-sEXPORTED_FUNCTIONS` doesn't work with `_emval_get_global`:

https://github.com/emscripten-core/emscripten/blob/0649692b8ab17bac09262754b68cbf6aff297963/system/include/emscripten/val.h#L75

I get the following:
```
emcc: error: undefined exported symbol: "_emval_get_global" [-Wundefined] [-Werror]
```
Regardless of whether I prepend an underscore or not.

**Reply by olokobayusuf** [reply]
On a somewhat separate note, I'm thinking of opening a PR because the `-Wl` flag ignores quotes and simply splits all commas. As such, the following errors:
```
emcc ... -Wl,--features=mutable-globals,sign-ext
```
And quotes don't work either. Relevant source with the culprit:

https://github.com/emscripten-core/emscripten/blob/0649692b8ab17bac09262754b68cbf6aff297963/emcc.py#L779-L785

Should I open a PR for this?

**Reply by sbc100** [reply]
`_emval_get_global` is a little special because its a JS symbol.  If its not used by you code you currently also need to add those to `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE`.    Its not great, but that is the way things are setup today.  So for example this command works:

```
$ emcc -sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE=_emval_get_global -sEXPORTED_FUNCTIONS=__emval_get_global,_main test/hello_world.c -lembind
```

Not that unlike `EXPORTED_FUNCTIONS`, `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE` *doesn't* use leading underscores to mangle the symbols.  I don't have a great rational for this other than its a historical artifact.

We have talked about having `EXPORTED_FUNCTIONS` work for JS library symbols even when they are not otherwise includes.   We have also discussed a completely new `-sEXPORTS` setting to replace all of the current ones (#8380) and removing the underscore mangling (#12844).  

You could also instead do `-sINCLUDE_FULL_LIBRARY` (which includes all JS library functions).

**Reply by olokobayusuf** [reply]
Integrated this and ran some more tests. I've pinpointed why my code still doesn't work, hopefully you can shed some light:

1. First, I'm currently using Emscripten 3.1.8 but the version doesn't quite matter for the mechanics.
2. Your suggestion on using `-sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE` is spot on; the necessary JS library function from `embind` is now available in the module.
3. My side module gets loaded at runtime using dynamic linking, and needs to use aforementioned JS library function from the DCE'd main module. From the dynamic linking implementation, the `WebAssembly.Module` is created with a set of `imports` from the main module (providing things like cstdlib and other runtime methods). The issue is that **there doesn't seem to be a way to add a symbol into `asmLibraryArg` unless it is explicitly used by the main module**.

In 3.1.8, `asmLibraryArg` is emitted by `create_module`:

https://github.com/emscripten-core/emscripten/blob/3ff7eff373d31d8c0179895165462019221f192e/emscripten.py#L837-L842

And `sending` is in turn created from the compiled WASM binary metadata (I presume this comes from the output of `wasm-ld` without any modifications):

https://github.com/emscripten-core/emscripten/blob/3ff7eff373d31d8c0179895165462019221f192e/emscripten.py#L719-L724

And metadata extraction is implemented here:

https://github.com/emscripten-core/emscripten/blob/3ff7eff373d31d8c0179895165462019221f192e/tools/extract_metadata.py#L165-L176

Both `metadata["declares"]` and `metadata["globalImports"]` are populated from `webassembly.Module.get_imports()`, so my question is what's missing? How come if I add the following line of code, `_emval_get_global` is properly declared as an import in my main module:
```cpp
auto someGlobal = emscripten::val::global("window");
```
But using combinations of `sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE`, `-sEXPORTED_FUNCTIONS`, `-Wl,-u_emval_get_global`, and `-Wl,--export-if-defined=__emval_get_global` all don't work?

**Reply by sbc100** [reply]
The symbols available to the side module (at least back in 3.18) are not only the ones in `asmLibraryArg`, but also any symbols exported on the `Module` object.  See `resolveGlobalSymbol`.

However, a lot of this code as been updated in recent versions of emscripten.  Is there any reason why you can't update to a more recent version.  3.1.8 is fairly old at this point.

Also, I'm not sure you answer the question: "Are you able you pass the side modules are augments when linking the main module?".

**Reply by olokobayusuf** [reply]
> The symbols available to the side module (at least back in 3.18) are not only the ones in asmLibraryArg, but also any symbols exported on the Module object. See resolveGlobalSymbol.

Good catch! I'm getting very close to resolving this. `resolveGlobalSymbol` has fallbacks to search for the missing symbol, in this case `symName` is `_emval_get_global` as an example:

https://github.com/emscripten-core/emscripten/blob/3ff7eff373d31d8c0179895165462019221f192e/src/library_dylink.js#L10-L30

`asmjsMangle(symName)` in this case adds a leading underscore, so we search for `Module["__emval_get_global"]`. The problem though that for whatever reason, I can't get the `__emval_get_global` JS function (which exists thanks to `-sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE`) to get exported into `Module` with `-sEXPORTED_RUNTIME_METHODS`. Using `-sEXPORTED_FUNCTIONS=__emval_get_global` doesn't help.

An interesting thing I noticed is this declaration in the resulting JS framework code:
```
unexportedRuntimeFunction('emval_get_global', false);
```
The `embind` JS library defines both `emval_get_global` and `__emval_get_global`, with the latter depending on the former. I can get `emval_get_global` added to `Module` using `-sEXPORTED_RUNTIME_METHODS`, but the same doesn't apply for `__emval_get_global`.

**Reply by olokobayusuf** [reply]
> However, a lot of this code as been updated in recent versions of emscripten. Is there any reason why you can't update to a more recent version. 3.1.8 is fairly old at this point.

We're currently adding support for Unity 2022, which uses Emscripten 3.1.8. Later versions of Unity have bumped their Emscripten version, but we need to support Unity versions all the way back to 2022.

> Also, I'm not sure you answer the question: "Are you able you pass the side modules are augments when linking the main module?".

Oh sorry. We can't provide side modules when compiling the main module. We have a pretty advanced use case: we're building a cross-platform Python function runtime, where we transpile and cross-compile Python functions then ship these compiled binaries to users' devices at runtime. To power all this, we exploit a ton of dynamic linking across Android, iOS, macOS, Windows, and WebAssembly. [Check us out](https://fxn.ai).

**Reply by sbc100** [reply]
I assume this is because you don't know that total set of possible side modules that you might load?  In which case, one shortcut to getting all the symbols you need might be to build a "sample"/"superset" side module that does a superset of things that you want to enable (.e.g it uses embind) and then link against that?

**Comment by sbc100** [comment]
> Should I open a PR for this?

Sure that sounds reasonable.

**Comment by sbc100** [comment]
> `asmjsMangle(symName)` in this case adds a leading underscore, so we search for `Module["__emval_get_global"]`. The problem though that for whatever reason, I can't get the `__emval_get_global` JS function (which exists thanks to `-sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE`) to get exported into `Module` with `-sEXPORTED_RUNTIME_METHODS`. Using `-sEXPORTED_FUNCTIONS=__emval_get_global` doesn't help.

When I try this on the (latest emscripten version at least) it what you want:

```
$ emcc test/hello_world.c -lembind -sMAIN_MODULE=2 -sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE=_emval_get_global -sEXPORTED_FUNCTIONS=__emval_get_global,_main
$ grep __emval_get_global a.out.js
...
  Module['__emval_get_global'] = __emval_get_global;
...
```

Does this not work for you?   It sounds like add `__emval_get_global` to `EXPORTED_FUNCTIONS` does not work.. but if a you try to add a symbol that desn't exist to that list then you would get an error.  e.g.

```
$ ./emcc ~/test/hello.c -lembind -sMAIN_MODULE=2 -sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE=_emval_get_global -sEXPORTED_FUNCTIONS=missing,_main
emcc: error: undefined exported symbol: "missing" [-Wundefined] [-Werror]
```

**Reply by sbc100** [reply]
It also seems to work for me with 3.1.8:

```
$ emcc ~/test/hello.c -lembind -sMAIN_MODULE=2 -sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE=_emval_get_global -sEXPORTED_FUNCTIONS=__emval_get_global,_main
$  grep __emval_get_global a.out.js
  function __emval_get_global(name) {
  Module["__emval_get_global"] = __emval_get_global;
  __emval_get_global.sig = 'ii';
```

**Reply by olokobayusuf** [reply]
No clue here, but I decided to include the symbols through source code (write nop code that uses the required symbols) instead of trying to do this through compiler and linker flags.

Now, I have a separate issue that is unrelated but since I have your attention I thought to go ahead and ask asnyway (lmk if I should create a separate discussion, thanks):

I am trying to use `emscripten::val` without RTTI (and the required`EMSCRIPTEN_HAS_UNBOUND_TYPE_NAMES 0` definition) but at runtime I get an `unbound type: emsc` error.  You've previously worked on this ([here](https://github.com/emscripten-core/emscripten/commit/685c627a9186b38869659089c74ae326b3d770cd)), and fixed it ([here](https://github.com/emscripten-core/emscripten/pull/9261)) but I'm struggling to understand how your fix actually works.

I think I'm facing this bug because of my custom setup: I have a main module being compiled with `-lembind -fno-rtti`, and I'm trying to use `emscripten::val::as<T>` in a side module that gets dynamically linked into the main module at runtime. I believe that based on the current implementation of `emscripten::val` (and the underlying `wire` impl), it uses a sentinel value to tell apart types at build time when `-fno-rtti` is used (i.e. in place of using `typeid` when using RTTI). But with a dynamically-linked side module, the type indices don't exist, so I get an invalid placeholder type `emsc`. Is my intuition correct?

If my intuition is correct, I can imagine a workaround where I patch `requireRegisteredType` JS library implementation (`embind.js`) to find the correct type. The issue though is that I can't do a type name search through `registeredType` because `getTypeName` doesn't seem to work when `EMSCRIPTEN_HAS_UNBOUND_TYPE_NAMES == 0`:

https://github.com/emscripten-core/emscripten/blob/9eff02bc816c50ab0e3b70a3bd5b72a8dc2893a2/system/lib/embind/bind.cpp#L45-L49

What do you think?

**Reply by olokobayusuf** [reply]
[This is the workaround I'm referring to](https://github.com/emscripten-core/emscripten/pull/18418), which would only work with RTTI.
